core/src/main/scala/org/apache/spark/deploy/worker/WorkerWatcher.scala

```diff
@@ -53,10 +53,9 @@ private[spark] class WorkerWatcher(
   private def exitNonZero() =
     if (isTesting) {
       isShutDown = true
-    } else {
-      ThreadUtils.awaitResult(Future {
-        System.exit(-1)
-      }, 5.seconds)
+    } else if (isChildProcessStopping.compareAndSet(false, true)) {
```

Comment:
We will need to modify exit behavior in `CoarseGrainedExecutorBackend` also to ensure atomic compare + set to prevent races, right ? (in `onDisconnected`)

Reply:
It could prevent but in the case of `StopExecutor` and `Shutdown`, the executor will exit the JVM anyway. So it doesn't matter.